packages/frontend/core/src/modules/open-in-app/utils.ts

```diff
@@ -0,0 +1,21 @@
+import { channelToScheme } from './constant';
+
+// return an AFFiNE app's url to be opened in desktop app
+export const getOpenUrlInDesktopAppLink = (
+  url: string,
+  newTab = false,
+  scheme = channelToScheme[BUILD_CONFIG.appBuildType]
+) => {
+  if (!scheme) {
+    return null;
+  }
+
+  const urlObject = new URL(url);
+  const params = urlObject.searchParams;
+
+  if (newTab) {
+    params.set('new-tab', '1');
+  }
+
+  return `${scheme}://${urlObject.hostname}${urlObject.pathname}?${params.toString()}#${urlObject.hash}`;
```

Comment:
Is it unnecessary to handle `port` | `empty search params` | `empty hash`?

Reply:
port is part of hostname;
if search params is empty we will have:
`affine://localhost:8080?#ssss`, which is a valid case

Reply:
![CleanShot 2024-10-25 at 09.58.18@2x.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/LakojjjzZNf6ogjOVwKE/e5a705fb-dd2d-4304-94e2-a36f88f94a0a.png)

🤔 `port` is included in `host` but not `hostname`?

Reply:
indeed. my bad. thanks for pointing out!